### PR TITLE
[LWG 5] P2770R0 Stashing stashing iterators for proper flattening

### LINKED ISSUE #6095: [2023-02 LWG Motion 5] P2770R0 Stashing stashing iterators for proper flattening

### PR DESCRIPTION
**Pull request by burblebee**
Fixes NB US 61-126 (C++23 CD).

Fixes #6095.
Fixes cplusplus/nbballot#539

### CODE REVIEW

File: source/ranges.tex [hunk 1]
```diff
@@ -7149,14 +7230,14 @@
 Equivalent to:
 \begin{codeblock}
 if constexpr (@\exposid{ref-is-glvalue}@)
-  return *x;
+  return *@\exposid{as-lvalue}@(@\exposid{outer}@());
```
Comment (JohelEGP) [inline source/ranges.tex:7233]:
```suggestion
  return @\exposid{as-lvalue}@(*@\exposid{outer}@());
```

File: source/ranges.tex [hunk 2]
```diff
@@ -7165,7 +7246,7 @@
...
```
Comment (JohelEGP) [inline source/ranges.tex:7249]:
```suggestion
  return @\exposid{as-lvalue}@(*@\exposid{outer}@());
```

File: source/ranges.tex [hunk 3]
```diff
@@ -7207,19 +7286,22 @@
 \end{itemdescr}
 
 \begin{itemdecl}
-constexpr @\exposid{iterator}@(@\exposid{Parent}@& parent, iterator_t<@\exposid{Base}@> outer);
+constexpr @\exposid{iterator}@(@\exposid{Parent}@& parent, @\exposid{OuterIter}@ outer)
+  requires @\libconcept{forward_range}@<@\exposid{Base}@>;
+constexpr explicit @\exposid{iterator}@(@\exposid{Parent}@& parent)
+  requires (!@\libconcept{forward_range}@<@\exposid{Base}@>);;
```
Comment (JohelEGP) [inline source/ranges.tex:7292]:
```suggestion
  requires (!@\libconcept{forward_range}@<@\exposid{Base}@>);
```

**Review by JohelEGP — CHANGES_REQUESTED** [review]

File: source/ranges.tex [hunk 4]
```diff
@@ -7014,12 +7077,18 @@
     static constexpr bool @\exposid{ref-is-glvalue}@ = is_reference_v<@\exposid{InnerBase}@>;   // \expos
 
     @\exposid{Parent}@* @\exposid{parent_}@ = nullptr;                                          // \expos
-    @\exposid{OuterIter}@ @\exposid{outer_it_}@ = @\exposid{OuterIter}@();                                  // \expos
+    @\exposid{OuterIter}@ @\exposid{outer_it_}@ = @\exposid{OuterIter}@();                          // \expos, present only
+                                                                // if \tcode{\exposid{Base}} models \libconcept{forward_range}
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/ranges.tex:7081]: FWIW, we use this formatting elsewhere in comments like this, but you're right, that's no reason to replicate that here just for the sake of consistency :). Fixed.

File: source/ranges.tex (hunk 2 above)
Comment (burblebee) [reply source/ranges.tex:7249]: Thanks!

File: source/ranges.tex (hunk 1 above)
Comment (burblebee) [reply source/ranges.tex:7233]: thanks!

File: source/ranges.tex (hunk 4 above)
[the remark this answers is not in the document]
Comment (JohelEGP) [reply source/ranges.tex]: A check for this should be easy @jensmaurer.

**Review by JohelEGP — APPROVED** [review]

**Pull request merged**